base/abstractarray.jl

```diff
@@ -86,8 +128,48 @@ linearindexing(A::AbstractArray, B::AbstractArray...) = linearindexing(linearind
 linearindexing(::LinearFast, ::LinearFast) = LinearFast()
 linearindexing(::LinearIndexing, ::LinearIndexing) = LinearSlow()
 
+abstract IndicesBehavior
+immutable IndicesStartAt1  <: IndicesBehavior end   # indices 1:size(A,d)
+immutable IndicesUnitRange <: IndicesBehavior end   # arb UnitRange indices
+immutable IndicesList      <: IndicesBehavior end   # indices like (:cat, :dog, :mouse)
+
+indicesbehavior(A::AbstractArray) = indicesbehavior(typeof(A))
+indicesbehavior{T<:AbstractArray}(::Type{T}) = IndicesStartAt1()
+indicesbehavior(::Number) = IndicesStartAt1()
+
+abstract IndicesPerformance
+immutable IndicesFast1D <: IndicesPerformance end  # indices(A, d) is fast
+immutable IndicesSlow1D <: IndicesPerformance end  # indices(A) is better than indices(A,d)
+
+indicesperformance(A::AbstractArray) = indicesperformance(typeof(A))
+indicesperformance{T<:AbstractArray}(::Type{T}) = IndicesFast1D()
+
+"""
+    shape(A)
+
+Returns a tuple specifying the "shape" of array `A`. For arrays with
+conventional indexing (indices start at 1), this is equivalent to
+`size(A)`; otherwise it is equivalent to `incides(A)`.
+"""
```

Comment:
If you want one or the other you have `size` and `indices`. `shape` is intended for cases where you'd like to be able to fall back to "conventional" array types, but don't want anything bad to happen if you can't do that.

Here's an example:
- `similar(::Array, (3,5))` should allocate an array of size 3x5
- `similar(::Array, (0:2, -2:2))` should allocate an `OffsetArray{_,_,Array}`.
- so to preserve type-stability, `similar(::Array, (1:3, 1:5))` should allocate an `OffsetArray{_,_,Array}`, even though you don't actually need to.

You'd have the 3rd happening if you called `similar(A, indices(B))`, but you get the first if you call `similar(A, shape(B))`.

Let me know if that's enough justification for `shape`. If not, I can change.